### PR TITLE
pybricks/connections/pybricks: add connection context manager

### PR DESCRIPTION
**Pull request by Novakasa**
added ``PybricksHub.connection(device)`` async context manager to be used like this:
```python
    dev = await find_device()
    async with hub.connection(dev):
        await hub.run("hub_messages.py")
```
In master, you would write:
```python
    dev = await find_device()
    try:
        await hub.connect(dev)
        await hub.run("hub_messages.py")
    finally:
        await hub.disconnect()
```

### CODE REVIEW

File: pybricksdev/connections/pybricks.py [hunk 1]
```diff
@@ -175,6 +176,19 @@ def pybricks_service_handler(self, _: int, data: bytes) -> None:
...
+        """Context manager to handle connect and disconnect for a device discovered
+        with :meth:`pybricksdev.ble.find_device`. To be used in an ``async with``
+        statement.
+
+        Args:
+            device: The device to connect to.
+        """
+        await self.connect(device)
+        yield
+        await self.disconnect()
```
Comment (dlech) [inline pybricksdev/connections/pybricks.py:190]: Does this need to be:
```suggestion
        await self.connect(device)
        try:
            yield
        finally:
            await self.disconnect()
```

...to ensure that disconnect is always called even if the code in the `async with` block throws an exception?

Comment (Novakasa) [reply pybricksdev/connections/pybricks.py:190]: Yeah you're right. I didn't know how this decorator worked exactly.

**Comment by Novakasa** [comment]
When testing for exceptions thrown during `connect()`, I noticed a missing `await` in the cleanup code there, so I added a fix.

**Comment by dlech** [comment]
While I am a big fan of context managers, I think with the current design of our API, there are many more cases where we would want a context manager, e.g. subscriptions to observables, stopping the user program, etc.

So I think were going to say "thanks, but no thanks" on this one and instead recommend using `contextlib.AsyncExitStack` which we probably should be using anyway to avoid lots of nested context managers.

Here is an example that will stop the user program and disconnect the hub when pressing CTRL+C to break out of the infinite loop.

```python
import asyncio
import contextlib

from pybricksdev.connections.pybricks import PybricksHub
from pybricksdev.ble import find_device


async def main():
    async with contextlib.AsyncExitStack() as stack:
        hub = PybricksHub()

        device = await find_device()

        await hub.connect(device)
        stack.push_async_callback(hub.disconnect)

        await hub.run("hub_test.py", wait=False, print_output=False)
        stack.push_async_callback(hub.stop_user_program)

        while True:
            await hub.write_line("x")
            line = await hub.read_line()
            print(line)


asyncio.run(main())
```

`hub_test.py`

```python
while True:
    line = input()
    print("echo:", line)
```

**Pull request closed without merging**